File: generated_code/py/program_2_2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        # Create a dummy head to simplify the linked list construction logic
        dummy_head = ListNode(0)
        curr = dummy_head
        carry = 0
        
        # Loop through both lists until both are exhausted and no carry remains
        while l1 or l2 or carry:
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0
            
            # Calculate the total sum for the current position
            total = val1 + val2 + carry
            carry = total // 10
            
            # Create a new node with the single-digit value
            curr.next = ListNode(total % 10)
            curr = curr.next
            
            # Move to the next nodes if they exist
            if l1: l1 = l1.next
            if l2: l2 = l2.next
                
        # Return the actual head of the resulting list (skipping the dummy node)
        return dummy_head.next

File: generated_code/cpp/program_2_2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 * int val;
 * ListNode *next;
 * ListNode() : val(0), next(nullptr) {}
 * ListNode(int x) : val(x), next(nullptr) {}
 * ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        // Use a dummy head to simplify building the result linked list.
        ListNode dummy(0);
        ListNode* tail = &dummy;
        int carry = 0;

        // Iterate while there are nodes remaining in either list or there is a carry.
        // Time Complexity: O(max(N, M)) where N and M are lengths of l1 and l2.
        // Space Complexity: O(1) auxiliary space (excluding the output list).
        while (l1 != nullptr || l2 != nullptr || carry > 0) {
            int sum = carry;

            if (l1 != nullptr) {
                sum += l1->val;
                l1 = l1->next;
            }
            if (l2 != nullptr) {
                sum += l2->val;
                l2 = l2->next;
            }

            carry = sum / 10;
            tail->next = new ListNode(sum % 10);
            tail = tail->next;
        }

        return dummy.next;
    }
};